Avoid hardcoding paths in Profile and Sys

nalimilan opened the pull request:
Current code fails when using custom paths. Save directory where Julia source is stored during build in `SOURCEDIR` instead of assuming it can be computed from `BINDIR` (new name chosen to avoid confusing with `build_dir` which is different). Use `DATAROOTDIR` and `DATAROOT`  instead of hardcoding `usr/share/` and `share/`.

Fix/continuation of #56601, #56627.

nalimilan committed:
Avoid hardcoding paths in Profile and Sys

Current code fails when using custom paths. Save directory where Julia source
is stored during build in `SOURCEDIR` instead of assuming it can be computed
from `BINDIR` (new name chosen to avoid confusing with `build_dir` which is different).
Use `DATAROOTDIR` and `DATAROOT`  instead of hardcoding `usr/share/` and `share/`.

Fix/continuation of #56601, #56627.

nalimilan commented on base/Makefile:
```diff
@@ -71,6 +71,7 @@ endif
 	@printf "%s\n" "const PRIVATE_LIBDIR = "$(call shell_escape,$(call julia_escape,$(call normalize_path,$(private_libdir_rel)))) >> $@
 	@printf "%s\n" "const PRIVATE_LIBEXECDIR = "$(call shell_escape,$(call julia_escape,$(call normalize_path,$(private_libexecdir_rel)))) >> $@
 	@printf "%s\n" "const INCLUDEDIR = "$(call shell_escape,$(call julia_escape,$(call normalize_path,$(includedir_rel)))) >> $@
+	@printf "%s\n" "const SOURCEDIR = "$(call shell_escape,$(call julia_escape,$(call normalize_path,$(JULIAHOME)))) >> $@
```
AFAICT this new constant is needed as paths to source files are those at build time. `BUILD_ROOT_PATH` which was defined below was incorrect in claiming that it pointed to the build-time source path as it was based on `BINDIR`, yet there's no guarantee that the julia executable during build time is next to the source dir.

nalimilan committed:
Windows fixes

Failure for `@juliasrc`: SOURCEDIR was "/c/workdir" (giving "\\c\\workdir") on Windows builder,
while "C:\\workdir" was expected.
Failure for `@Compiler`: missing ending slash meant that remainder was an absolute path.
Minor cleanups.

nalimilan committed:
Another attempt

nalimilan commented:
CI failure on test x86_64-linux-gnuassertrr seems unrelated.

IanButterworth committed:
Merge branch 'master' into nl/datarootdir

KristofferC commented:
I see fewer hardcoded paths so that is nice. Should the stuff in https://github.com/JuliaLang/julia/pull/57274 be updated as well?

nalimilan committed:
Do not hardcode "../share"

nalimilan commented:
I think it's correct as with the current options stdlib and Compiler will always live under `$(DATAROOTDIR)/julia`, but it can be written in a cleaner way I think. I've added a commit which does that and also replaced where `../share` is hardcoded.

This partly overlaps with https://github.com/JuliaLang/julia/pull/54353 which I'm trying to finish but I get some errors that I haven't been able to fix yet.

KristofferC reviewed (approved):
Nice, LGTM (assuming CI is ok)

nalimilan commented:
Thanks!
